docs/why-and-how-of-inline-directives.md

```diff
@@ -0,0 +1,369 @@
+# Why and How of Inline Directives
+
+This page explains four things:
+
+1. The primary optimization `purs-backend-es` does
+2. How inlining and inline directives help rewrite rules to optimize code
+3. The most cost-efficient methodology for defining inline directives that are useful
+4. Guiding principles as to where to define such inline directives
+
+## The Goal of Optimizations
+
+What's the difference between the following two expressions?
+
+```purs
+case1 :: forall a. FailureOrSuccess a -> Boolean
+case1 someComputation =
+  maybeToBoolean $ eitherToMaybe $ resultToEither someComputation
+  where
+  resultToEither :: FailureOrSuccess a -> Either String a
+  resultToEither arg = case  of
+    Failure -> Left "There was an error!"
+    Success a -> Right a
+
+  eitherToMaybe :: Either String a -> Maybe a
+  eitherToMaybe arg = case arg of
+    Left _ -> Nothing
+    Right a -> Just a
+
+  maybeToBoolean :: Maybe a -> Boolean
+  maybeToBoolean arg = case arg of
+    Nothing -> false
+    Just a -> true
+
+case2 :: forall a. FailureOrSuccess a -> Boolean
+case2 someComputation = case someComputation of
+  Failure -> false
+  Success _ -> true
+```
+
+While a developer would never intentionally write an expression like `case1`, such an expression may still arise if one has a long enough "chain" of functions spread across different modules.
+
+<details>
+<summary>For an example of such a "chain", expand this accordion</summary>
+
+```purs
+-- Module1.purs
+someComputation :: forall a. Int -> FailureOrSuccess a
+
+-- Module2.purs
+resultToEither :: FailureOrSuccess a -> Either String a
+resultToEither arg = case  of
+  Failure -> Left "There was an error!"
+  Success a -> Right a
+
+computationAsEither = resultToEither $ someComputation 4
+
+-- Module3.purs
+eitherToMaybe :: Either String a -> Maybe a
+eitherToMaybe arg = case arg of
+  Left _ -> Nothing
+  Right a -> Just a
+
+computationAsMaybe = map doSomethingElse computationAsEither
+
+-- Module4.purs
+maybeToBoolean :: Maybe a -> Boolean
+maybeToBoolean arg = case arg of
+  Nothing -> false
+  Just a -> true
+
+-- Module5.purs
+foo = do
+  let x = maybeToBoolean computationAsMaybe
+  ...
+```
+
+</details>
+
+At the end of the day, evaluating `case1` and `case2` produce the same boolean value, but `case1` produces and then consumes 2 intermediate data structures unnecessarily: `Either` and `Maybe`.
+
+**To summarize, the goal of our optimizations is to remove these unneeded intermediate data structures.** As a result, the code is more performant because there's less purely-overhead work for the computer to do at runtime.
+
+## How Code Gets Optimized
+
+### Primitive Data Flow: Production and Consumption
+
+How does `purs-backend-es` know when unneeded intermediate data structures are being used and when it is safe to remove them? These usages arise when primitive data is "produced" and then immediately "consumed".
```

Comment:
I'm not sure the producer/consumer nomenclature is quite appropriate here. More commonly you'll see "constructor" and "eliminator". It's the same duality though. That is, `Right` is the constructor, and `case` is the eliminator.

Reply:
An accessor being sugar for a primitive `case`.

```
getFoo = case _ of
  { foo } -> foo
```